### ISSUE TITLE
cipher failures in master 2019-08-03

### ISSUE THREAD

**Issue by petrovr** [openssl/openssl#9524]
With recent code from master branch PKIX-SSH rekey tes fail when is used aes{128|256}-gcm.
Last know good version is from 2019-07-13.

Issue is not investigated yet, i.e. dunno why OpenSSL code pass internal tests.

**Comment by mattcaswell** [comment]
Possibly related to #9512?

**Comment by petrovr** [comment]
mattcaswell wrote:
> Possibly related to #9512?

Above issue is not related. Patch does not change anything.

Issue is still not investigated yet. In client failure logs is printed 
'padding error: need 260 block 16 mod 4' ..

Roumen

**Comment by slontis** [comment]
Probably an edge case that is not tested.
Let me know if you have any more info.

**Comment by petrovr** [comment]
Same situation as of today.
Simplified test case is just following command: ./ssh o Ciphers=aes128-gcm@openssh.com ...
Unfortunately build of pkix-ssh against master is partially blocked by #9522 .

Failed function call is this one https://gitlab.com/secsh/pkixssh/blob/master/cipher.c#L479 .

Now is easy to reconstruct sample:
- get the sample code from https://github.com/openssl/openssl/blob/master/demos/evp/aesgcm.c
- update some variables:
IV:
```c
static const unsigned char gcm_iv[] = {
    0x99
};
```
In consequence ciphertext:
```c
static const unsigned char gcm_ct[] = {
    0xf7, 0x26, 0x44, 0x13, 0xa8, 0x4c, 0x0e, 0x7c, 0xd5, 0x36, 0x86, 0x7e,
    0xb9, 0xf2, 0x17, 0x36
};
```
and tag:
```c
static const unsigned char gcm_tag[] = {
    0x54, 0x90, 0x40, 0x72, 0x52, 0x4d, 0x01, 0x31, 0x31, 0x13, 0xe8, 0x63,
    0x47, 0x75, 0x41, 0xe3
};
```
Remark: values are from 1.1.1 (Stable).

Build and run against master ....

**Comment by slontis** [comment]
I tried putting this data into evpcipt.txt and it reports that the ivlen is bad (as it should).
That demo program is very dodgy and does no error checking. Debugging the demo code showed that it gets an error when setting a bad iv length.

**Comment by petrovr** [comment]
Shane wrote:
> I tried putting this data into evpcipt.txt and it reports that the ivlen is bad (as it should).
> That demo program is very dodgy and does no error checking. Debugging the demo code showed that it gets an error when setting a bad iv length.
> 

Sorry using demo code was not good idea - does not match real case. 
Please Ignore it.

The line where code fail is correct. But writing test case is big challenge.

Please find attached "0003-INPROGRESS-use-ERR_raise.patch" that allows 
pkix-ssh code to be build against master.

Regards,
Roumen Petrov

**Comment by petrovr** [comment]
Download URL: 
https://roumenpetrov.info/tests/0003-INPROGRESS-use-ERR_raise.patch as 
github mail remove attached file.

**Comment by petrovr** [comment]
Hmm, pkix-ssh builds fine against master unless if OpenSSL is build with --api=3.0.0.
Remark: OpenSSL  build with --api=3.0.0 and error management impact is subject of another issues for instance #9756. PKIX-SSH will be updated when is clarified new error functionality.

I try to extract OpenSSL errors after failed function call but errors list is empty!
No error management in providers!? :(

**Comment by levitte** [comment]
> No error management in providers!? :(

Not entirely true. However, the providers currently rely entirely on error tables from the libcrypto code. 
This will change

**Comment by slontis** [comment]
I am a bit tied up with other work at the moment, I will try this out later in the week.

**Comment by slontis** [comment]
what should my config look like?
./configure --with-ssl-dir=openssl_extern/openssl

I get..
checking OpenSSL header version... 3.0.0 (OpenSSL 3.0.0-dev xx XXX xxxx)
checking OpenSSL library version... not found
configure: error: OpenSSL library not found.

**Comment by petrovr** [comment]
I guest that team members found what is missing with aes-gcm provider implementation.

**Comment by mattcaswell** [comment]
> I guest that team members found what is missing with aes-gcm provider implementation.

Does that mean this issue is fixed?

**Comment by slontis** [comment]
I got distracted by other things, so never got back to this one..
Only a few small changes have happened in gcm, so I dont know why it would be working now?
Is it working now?

**Comment by levitte** [comment]
It may have been fixed because #10106 is fixed...

**Comment by petrovr** [comment]
Richard Levitte wrote:
> It may have been fixed because #10106 is fixed...
> 

No.Restoring ctrl() will shows more precise failure - provider lack some 
GCM specific control commands.

I'm busy with other task ... I could provide later...

Roumen

**Comment by petrovr** [comment]
Let review result of:
$ grep -r EVP_CTRL_GCM crypto/ providers/ include/
```
crypto/evp/evp_enc.c:    case EVP_CTRL_GCM_SET_IV_FIXED:
crypto/evp/e_aria.c:    case EVP_CTRL_GCM_SET_IV_FIXED:
crypto/evp/e_aria.c:    case EVP_CTRL_GCM_IV_GEN:
crypto/evp/e_aria.c:    case EVP_CTRL_GCM_SET_IV_INV:
crypto/evp/e_aria.c:                            EVP_CTRL_GCM_IV_GEN : EVP_CTRL_GCM_SET_IV_INV,
crypto/evp/e_aes.c:    case EVP_CTRL_GCM_SET_IV_FIXED:
crypto/evp/e_aes.c:    case EVP_CTRL_GCM_IV_GEN:
crypto/evp/e_aes.c:    case EVP_CTRL_GCM_SET_IV_INV:
crypto/evp/e_aes.c:    if (EVP_CIPHER_CTX_ctrl(ctx, enc ? EVP_CTRL_GCM_IV_GEN
crypto/evp/e_aes.c:                                     : EVP_CTRL_GCM_SET_IV_INV,
crypto/evp/e_aes.c:    case EVP_CTRL_GCM_SET_IV_FIXED:
crypto/evp/e_aes.c:    case EVP_CTRL_GCM_IV_GEN:
crypto/evp/e_aes.c:    case EVP_CTRL_GCM_SET_IV_INV:
crypto/evp/e_aes.c:    if (EVP_CIPHER_CTX_ctrl(ctx, ctx->encrypt ? EVP_CTRL_GCM_IV_GEN
crypto/evp/e_aes.c:                                              : EVP_CTRL_GCM_SET_IV_INV,
include/openssl/evp.h:# define         EVP_CTRL_GCM_SET_IVLEN          EVP_CTRL_AEAD_SET_IVLEN
include/openssl/evp.h:# define         EVP_CTRL_GCM_GET_TAG            EVP_CTRL_AEAD_GET_TAG
include/openssl/evp.h:# define         EVP_CTRL_GCM_SET_TAG            EVP_CTRL_AEAD_SET_TAG
include/openssl/evp.h:# define         EVP_CTRL_GCM_SET_IV_FIXED       EVP_CTRL_AEAD_SET_IV_FIXED
include/openssl/evp.h:# define         EVP_CTRL_GCM_IV_GEN             0x13
include/openssl/evp.h:# define         EVP_CTRL_GCM_SET_IV_INV         0x18
```

So in evp_enc.c EVP_CTRL_GCM_SET_IV_FIXED could be replaced with EVP_CTRL_AEAD_SET_IV_FIXED, right?

Question is how provider implement EVP_CTRL_GCM_IV_GEN and EVP_CTRL_GCM_SET_IV_INV.

Secure shell code uses EVP_CTRL_GCM_IV_GEN .

**Comment by slontis** [comment]
Ahh yeah ok..
EVP_CIPHER_CTX_ctrl() is missing some translations that it needs to do..
I will add them today..

**Comment by slontis** [comment]
See PR #10173

**Comment by petrovr** [comment]
This issue is for aes*gcm but now master has other failures... :(
For instance try-ciphers test hang on first test:
<pre>
=== run test try-ciphers.sh ...
.../regress/try-ciphers.sh: warning: line 8: `...` obsolete, use $(...)
.../regress/try-ciphers.sh: warning: line 10: `...` obsolete, use $(...)
.../regress/try-ciphers.sh: warning: line 25: `...` obsolete, use $(...)
<b>test try ciphers: cipher 3des-cbc mac hmac-sha1</b>
</pre>

**Comment by slontis** [comment]
>So in evp_enc.c EVP_CTRL_GCM_SET_IV_FIXED could be replaced with EVP_CTRL_AEAD_SET_IV_FIXED, right?



They are defined as the same thing - so either one will work.. And yes ccm handles it .
cipher_ccm uses OSSL_CIPHER_PARAM_AEAD_TLS1_IV_FIXED.

EVP_CIPHER_CTX_ctrl() does the translation from a CTRL to a OSSL_CIPHER_PARAM

**Comment by petrovr** [comment]
Shane wrote:
>> So in evp_enc.c EVP_CTRL_GCM_SET_IV_FIXED could be replaced with EVP_CTRL_AEAD_SET_IV_FIXED, right?
> 
> 
> 
> They are defined as the same thing - so either one will work.. And yes ccm handles it .
> cipher_ccm uses OSSL_CIPHER_PARAM_AEAD_TLS1_IV_FIXED.
> 
> EVP_CIPHER_CTX_ctrl() does the translation from a CTRL to a OSSL_CIPHER_PARAM
> 

Let FOO is equal to COMMON and BAR also is equal to COMMON.
So based on above comment developer could use in code randomly either 
FOO, or BAR, or COMMON. It will work.

This is not model that I will use. I would like such style to be 
discouraged.


Regards,
Roumen Petrov

P.S. for protocol COMMON is already used:
$ grep -n EVP_CTRL_AEAD */*/evp*.c
crypto/evp/evp_enc.c:1099:    case EVP_CTRL_AEAD_SET_IVLEN:
crypto/evp/evp_enc.c:1117:    case EVP_CTRL_AEAD_GET_TAG:
crypto/evp/evp_enc.c:1119:    case EVP_CTRL_AEAD_SET_TAG:
crypto/evp/evp_enc.c:1123:    case EVP_CTRL_AEAD_TLS1_AAD:

**Comment by petrovr** [comment]
I wonder how to proceed with this issue. Look like that gcm is corrected (see above for number). Unfortunately  correction cannot be tested in third party software as the test fail in another cipher before gcm :(

Dunno why OpenSSL test suite do not catch those failures.

**Comment by levitte** [comment]
Is this still an issue?

**Comment by petrovr** [comment]
Richard Levitte wrote:
> Is this still an issue?
> 

It seems to aes{128|256}-gcm is fixed in another issue (PR). I have just 
one remark - code to use more common *AEAD* *(1) macros instead *GSM* 
(2) . Code will work in both cases but (1) is more readable.

Unfortunately cipher test start to fail for another cipher ( see 
previous comment 
https://github.com/openssl/openssl/issues/9524#issuecomment-541802481 ).

I'm not sure when I could test current master.

Roumen

**Comment by petrovr** [comment]
Ciphers tests still hang with master 2019-12-14 .

**Comment by petrovr** [comment]
As expected no change.
For functional point of view issue does not differ from previous report (#8895) : broken serialization and deserialization.
Case is not covered by OpenSSL regression test. Simple check was used to test previous report. Now I have no idea what exacty is broken.
For sure if after some encryption/description operation, serialization and deserialization of chiper&mac, next opeation will fail.

**Comment by levitte** [comment]
@petrovr, I'm not familiar enough with pkix-ssh to figure out how to start looking more deeply into the code to figure out what's wrong and why, and it's not easy to see clearly what programs are run when and how.  If you have an easy recipe for me to run one test program in gdb, that would be awesome.

**Comment by petrovr** [comment]
Richard Levitte wrote:
> @petrovr, I'm not familiar enough with pkix-ssh to figure out how to start looking more deeply into the code to figure out what's wrong and why, and it's not easy to see clearly what programs are run when and how.  If you have an easy recipe for me to run one test program in gdb, that would be awesome.
> 
> 

PKIX-SSH (unlike another implementation) allows "privilege separation" ( 
https://github.com/openssl/openssl/issues/8895#issuecomment-509313858 ) 
to be disabled from configuration. If is disabled in cipher tests case 
"hang" disappear. In this case we will see failures with gsm (out of 
scope to this post).

So what this mean - serialization and deserialization error .

After authentication unprivileged process sends "packet state" (1) 
function newkeys_to_blob from  packet.c and exists.

```c
	if ((r = cipher_get_keyiv(cc, enc->iv, enc->iv_len)) != 0)
		return r;
...
	if ((r = sshbuf_put_cstring(b, enc->name)) != 0 ||
	    (r = sshbuf_put_u32(b, enc->enabled)) != 0 ||
	    (r = sshbuf_put_u32(b, enc->block_size)) != 0 ||
	    (r = sshbuf_put_string(b, enc->key, enc->key_len)) != 0 ||
	    (r = sshbuf_put_string(b, enc->iv, enc->iv_len)) != 0)
```

Remark: Function cipher_get_keyiv calls EVP_CIPHER_CTX_* if "key" 
functionality is managed by OpenSSL code.


Daemon starts new child process and "packet state" is deserialized(2) 
with data from (1) - function newkeys_from_blob from packet.c.

```c
	if ((r = sshbuf_get_cstring(b, &enc->name, NULL)) != 0 ||
	    (r = sshbuf_get_u32(b, (u_int *)&enc->enabled)) != 0 ||
	    (r = sshbuf_get_u32(b, &enc->block_size)) != 0 ||
	    (r = sshbuf_get_string(b, &enc->key, &keylen)) != 0 ||
	    (r = sshbuf_get_string(b, &enc->iv, &ivlen)) != 0)
		goto out;
```

Same for mac if applicable.



Let see in previous sample in this comment - 
https://github.com/openssl/openssl/issues/8895#issuecomment-509319798 .
Currently with master issue is similar - IV was zero (issue #8895) and 
now dump is with original(initial) IV.

It seems to me all ciphers are impacted - the ssh test try-ciphers fail 
for all [cipher,mac] pairs (for some test hang, for other fail, for 
aes-gsm fix is in a separate issue).


Let me quote part of my previous comment "For sure if after some 
encryption/description operation, serialization and deserialization of 
chiper&mac, next operation will fail"

Let imagine following test:
0) producer and consumer are initialized with same keys iv and etc.
1) producer encrypts "sample text" and sent (store to file, etc.)
2) consumer decrypts text
3) consumer gets cipher state and initialize new cipher instance with 
the state.
4) producer encrypts "sample text" and sent (store to file, etc.)
5) consumer decrypts text using "new" cipher instance - expected is 
correct decryption.

Above test is similar to ssh processes.


Regards,
Roumen Petrov

**Comment by jwalch** [comment]
I am working on a project where I have to integrate both OpenSSL & OpenSSH and I have been experiencing a similar symptom to what it is most recently described here. I can SSH into my system if it's AESGCM, but not with AESCTR. I know I'm saying OpenSSH and not PKIX-SSH, but it has a lot of the same code and so far as things are described many of the same mechanisms (privilege separation, trying to marshal across the cipher state, etc.)

I've been trying to debug-by-printf (ugh) this situation from OpenSSH side, and I think I have some idea of what happened.

I added some tracing to the cipher_init() and cipher_crypt() routines, along with the serialization / de-serialization routines mentioned above by @petrovr. I discovered that cipher_init() was getting called with some IV, then cipher_crypt() would get called a handful of times, then the same IV as shown in cipher_init() eventually got marshaled across in the blob. At first I actually thought this is what I SHOULD be seeing. Now I am less certain.

In OpenSSH, there is an EVP_CIPHER_CTX_get_iv() routine. This overlaps with an OpenSSL routine of the same name, of course... so with the right defines / configuration it calls into the OpenSSL one instead. But looking at the OpenSSH routine is instructive...

```
int
EVP_CIPHER_CTX_get_iv(const EVP_CIPHER_CTX *ctx, unsigned char *iv, size_t len)
{
	if (ctx == NULL)
		return 0;
	if (EVP_CIPHER_CTX_iv_length(ctx) < 0)
		return 0;
	if (len != (size_t)EVP_CIPHER_CTX_iv_length(ctx))
		return 0;
	if (len > EVP_MAX_IV_LENGTH)
		return 0; /* sanity check; shouldn't happen */
	/*
	 * Skip the memcpy entirely when the requested IV length is zero,
	 * since the iv pointer may be NULL or invalid.
	 */
	if (len != 0) {
		if (iv == NULL)
			return 0;
# ifdef HAVE_EVP_CIPHER_CTX_IV
		memcpy(iv, EVP_CIPHER_CTX_iv(ctx), len);
# else
		memcpy(iv, ctx->iv, len);
# endif /* HAVE_EVP_CIPHER_CTX_IV */
	}
	return 1;
}
```

It seems, at least to me, that the above is meant to get the running IV and not the original IV.

However, the OpenSSL routine by the same name introduced in 3.0.0 I *believe* is returning the original IV. The running IV is available in EVP_CIPHER_CTX_get_iv_state().

**Comment by jwalch** [comment]
Update: Switching the OpenSSH call to EVP_CIPHER_CTX_get_iv() instead to be EVP_CIPHER_CTX_get_iv_state() fixed the problem as observed in my setup.

**Comment by slontis** [comment]
@jwalch - what you are saying make sense to me. This is a real clash of namespaces here.
EVP_CIPHER_CTX_get_iv() in openssl is the original iv, so that is a problem.

**Comment by kaduk** [comment]
> @jwalch - what you are saying make sense to me. This is a real clash of namespaces here.
> EVP_CIPHER_CTX_get_iv() in openssl is the original iv, so that is a problem.

We still have a chance to rename ours before beta...

**Comment by jwalch** [comment]
> We still have a chance to rename ours before beta...

I thought of making that suggestion. On one hand, you CAN and it fixes the issue without any change being necessary on OpenSSH side. On the other hand, they sort of asked for it with the way they set up the detection of symbols and configuration defines... IMO, they really should not have assumed that a EVP_CIPHER_CTX_get_iv(), which didn't exist in 1.0.2 or 1.1.1 AFAIK, if eventually defined by OpenSSL would have the behavior they wanted. Seems like it would have been better to get a multiple definition error than to have this crazy run-time issue... IDK, maybe way back when or during development of some release OpenSSL had defined a EVP_CIPHER_CTX_get_iv() that did in fact return running IV...

**Comment by jwalch** [comment]
Oh, I think I've misread the history. It seems EVP_CIPHER_CTX_get_iv() is a LibreSSL construct. Go figure.

**Comment by paulidale** [comment]
@openssl/omc is a function name clash with LibreSSL a concern?

**Comment by mattcaswell** [comment]
> @openssl/omc is a function name clash with LibreSSL a concern?

We should make it easy to switch from LibreSSL to OpenSSL. So if we have the opportunity to the name things in a sensible way that allows that, I tend to think that we should.

**Comment by levitte** [comment]
Our previous (1.1.1) function names were `EVP_CIPHER_CTX_original_iv` for the original IV and `EVP_CIPHER_CTX_iv` for the running IV.  Maybe we should simply switch back to them...

**Comment by kaduk** [comment]
> Our previous (1.1.1) function names were `EVP_CIPHER_CTX_original_iv` for the original IV and `EVP_CIPHER_CTX_iv` for the running IV. Maybe we should simply switch back to them...

That would be a really bad idea, since the function signature is different.

> EDIT: ... or for using `get` more consistently, have `EVP_CIPHER_CTX_get_original_iv` for the original IV and `EVP_CIPHER_CTX_get_iv` for the running IV. Incidently, that would make them compatible with LibreSSL, as I understand it. This these are new names in 3.0, we have the freedom to adjust appropriately.

I would be okay with that.  IIRC I went with the `_get_iv`/`_get_iv_state` names because previous discussions had leaned strongly towards reserving the "iv" term for the actual *initial* vector

**Comment by levitte** [comment]
> ... previous discussions had leaned strongly towards reserving the "iv" term for the actual *initial* vector

Considering IV literally means "initialization vector", I hope that's not too surprising :wink:

That being said, I can see that the changed stance is more trouble than it's worth, so I don't mind changing it back (even though I gave it a good fight earlier)

**Comment by levitte** [comment]
Ah, something I forgot was the semantics.  The `const unsigned char *` returning functions are deprecated in favor of "fill my buffer" style functions (`EVP_CIPHER_CTX_get_iv` and `EVP_CIPHER_CTX_get_iv_state`), which I do think are a better fit for the future.

That gave me pause...

**Comment by kroeckx** [comment]
So should we just rename our _get_iv() to _get_initial_iv()?

**Comment by petrovr** [comment]
- A)

PKIX-SSH uses a model that ensures compatibility with different cryptographic library and versions.
Goal of "EVP compat" model is to uses most recent API.
For instance PKIX-SSH 2 years before OpenBSD managed secsh PKIX-SSH support 1.1.1 API.
Also model ensures compatibility with ancient OpenSSL releases like 0.9.7 with zero maintenance cost.

On other side OpenBSD still uses deprecated in 0.9.8 functionality and mechanically requires more recent releases of cryptographic library.

So @jwalch comment is not for this issue.
In particular case PKIX-SSH uses function cipher_get_keyiv uses EVP_CIPHER_CTX_iv (part of API 1.1).
As result @jwalch comment is applicable to this issue. It is up to OpenBSD to resolve failure.


- B)

PKIX-SSH fully support Android.
For instance you could get it from user applications: SecureBox with TermOnePlus as terminal or "SecureBox Pro" with integrated terminal and user interface.
Side effect is that privilege separation is still configuration option.
Privilege separation does not make sense in Android user application.

OpenBSD does drops configuration for privilege separation.


- C)

Serialisation issue impacts following regression tests: integrity, rekey and try-ciphers. Mostly identical between both secsh implementation.
PKIX-SSH fips-try-ciphers is out of scope - OpenBSD lack support and OpenSSL fips in master is different.

If PKIX-SSH regression tests are modified do not use privilege separation above tests pass.

This work-around is non recommended for production use.


- D)

It is clear that OpenSSL team fail to provide reliable code for more then one year. Also this issue shows that OpenSSL lack strong regression tests.

Status of 2020-08-28 (last tested, one year and 25 days after report):
  integrity: chacha20-poly1305@openssh.com fail, other ?
  rekey: lot of failures
  try-ciphers: ssh failed with mac hmac-sha1 cipher chacha20-poly1305@openssh.com
Dunno if chacha20-poly1305 is related to serialisation.


--- 
Please do not hijack issue. One day serialisation will be fixed.

I hope with added relevant tests.
Hint: for instance issue #8895 was to get right IV just after initialisation while this is more related to get IV after operation.

**Comment by levitte** [comment]
> So should we just rename our _get_iv() to _get_initial_iv()?

... or to `_get_original_iv()`, to reflect the old name?  (also "initial IV" sounds a bit like "CD disk" :wink:)

**Comment by petrovr** [comment]
As alredy pointed in my prevous comment function name clash is not appicable to this report.

**Comment by levitte** [comment]
> Hint: for instance issue #8895 was to get right IV just after initialisation while this is more related to get IV after operation.

Thanks for that hint

**Comment by jwalch** [comment]
> Please do not hijack issue.

**shrug**

No good deed goes unpunished, I suppose. Will file a new issue. Sorry I couldn't parse this bug report...

**Comment by t8m** [comment]
```
=== run test rekey.sh ...
client rekey KexAlgorithms=diffie-hellman-group1-sha1
client rekey KexAlgorithms=diffie-hellman-group14-sha1
client rekey KexAlgorithms=diffie-hellman-group14-sha256
client rekey KexAlgorithms=diffie-hellman-group16-sha512
client rekey KexAlgorithms=diffie-hellman-group18-sha512
client rekey KexAlgorithms=diffie-hellman-group-exchange-sha1
client rekey KexAlgorithms=diffie-hellman-group-exchange-sha256
client rekey KexAlgorithms=ecdh-sha2-nistp256
client rekey KexAlgorithms=ecdh-sha2-nistp384
client rekey KexAlgorithms=ecdh-sha2-nistp521
client rekey KexAlgorithms=curve25519-sha256
client rekey KexAlgorithms=curve25519-sha256@libssh.org
client rekey Ciphers=3des-cbc
client rekey Ciphers=aes128-cbc
client rekey Ciphers=aes192-cbc
client rekey Ciphers=aes256-cbc
client rekey Ciphers=aes128-ctr
client rekey Ciphers=aes192-ctr
client rekey Ciphers=aes256-ctr
client rekey Ciphers=aes128-gcm@openssh.com
client rekey Ciphers=aes256-gcm@openssh.com
client rekey Ciphers=chacha20-poly1305@openssh.com
client rekey MACs=hmac-sha1
client rekey MACs=hmac-sha1-96
client rekey MACs=hmac-sha2-256
client rekey MACs=hmac-sha2-512
client rekey MACs=hmac-md5
client rekey MACs=hmac-md5-96
client rekey MACs=umac-64@openssh.com
client rekey MACs=umac-128@openssh.com
client rekey MACs=hmac-sha1-etm@openssh.com
client rekey MACs=hmac-sha1-96-etm@openssh.com
client rekey MACs=hmac-sha2-256-etm@openssh.com
client rekey MACs=hmac-sha2-512-etm@openssh.com
client rekey MACs=hmac-md5-etm@openssh.com
client rekey MACs=hmac-md5-96-etm@openssh.com
client rekey MACs=umac-64-etm@openssh.com
client rekey MACs=umac-128-etm@openssh.com
client rekey aes128-gcm@openssh.com diffie-hellman-group1-sha1
client rekey aes128-gcm@openssh.com diffie-hellman-group14-sha1
client rekey aes128-gcm@openssh.com diffie-hellman-group14-sha256
client rekey aes128-gcm@openssh.com diffie-hellman-group16-sha512
client rekey aes128-gcm@openssh.com diffie-hellman-group18-sha512
client rekey aes128-gcm@openssh.com diffie-hellman-group-exchange-sha1
client rekey aes128-gcm@openssh.com diffie-hellman-group-exchange-sha256
client rekey aes128-gcm@openssh.com ecdh-sha2-nistp256
client rekey aes128-gcm@openssh.com ecdh-sha2-nistp384
client rekey aes128-gcm@openssh.com ecdh-sha2-nistp521
client rekey aes128-gcm@openssh.com curve25519-sha256
client rekey aes128-gcm@openssh.com curve25519-sha256@libssh.org
client rekey aes256-gcm@openssh.com diffie-hellman-group1-sha1
client rekey aes256-gcm@openssh.com diffie-hellman-group14-sha1
client rekey aes256-gcm@openssh.com diffie-hellman-group14-sha256
client rekey aes256-gcm@openssh.com diffie-hellman-group16-sha512
client rekey aes256-gcm@openssh.com diffie-hellman-group18-sha512
client rekey aes256-gcm@openssh.com diffie-hellman-group-exchange-sha1
client rekey aes256-gcm@openssh.com diffie-hellman-group-exchange-sha256
client rekey aes256-gcm@openssh.com ecdh-sha2-nistp256
client rekey aes256-gcm@openssh.com ecdh-sha2-nistp384
client rekey aes256-gcm@openssh.com ecdh-sha2-nistp521
client rekey aes256-gcm@openssh.com curve25519-sha256
client rekey aes256-gcm@openssh.com curve25519-sha256@libssh.org
client rekey chacha20-poly1305@openssh.com diffie-hellman-group1-sha1
client rekey chacha20-poly1305@openssh.com diffie-hellman-group14-sha1
client rekey chacha20-poly1305@openssh.com diffie-hellman-group14-sha256
client rekey chacha20-poly1305@openssh.com diffie-hellman-group16-sha512
client rekey chacha20-poly1305@openssh.com diffie-hellman-group18-sha512
client rekey chacha20-poly1305@openssh.com diffie-hellman-group-exchange-sha1
client rekey chacha20-poly1305@openssh.com diffie-hellman-group-exchange-sha256
client rekey chacha20-poly1305@openssh.com ecdh-sha2-nistp256
client rekey chacha20-poly1305@openssh.com ecdh-sha2-nistp384
client rekey chacha20-poly1305@openssh.com ecdh-sha2-nistp521
client rekey chacha20-poly1305@openssh.com curve25519-sha256
client rekey chacha20-poly1305@openssh.com curve25519-sha256@libssh.org
client rekeylimit 16
client rekeylimit 1k
client rekeylimit 128k
client rekeylimit 256k
client rekeylimit default 5
client rekeylimit default 10
client rekeylimit default 5 no data
client rekeylimit default 10 no data
server rekeylimit 16
server rekeylimit 1k
server rekeylimit 128k
server rekeylimit 256k
server rekeylimit default 5 no data
server rekeylimit default 10 no data
rekeylimit parsing
ok rekey
```

This is what I get when I run the pkixssh test against the current master. So apparently the original issue should be now fixed.

Unless you disagree @petrovr I'd recommend closing this issue and opening new issues if there are further issues that you see with OpenSSL master in regards to pkixssh tests.

**Comment by petrovr** [comment]
Hi,

Tomáš Mráz wrote:
> ```
> === run test rekey.sh ...
> client rekey KexAlgorithms=...
> ... ...
> rekeylimit parsing
> ok rekey
> ```

Good news.

> 
> This is what I get when I run the pkixssh test against the current master. So apparently the original issue should be now fixed.
> 
> Unless you disagree @petrovr I'd recommend closing this issue and opening new issues if there are further issues that you see with OpenSSL master in regards to pkixssh tests.
> 
> 

Does team plan to add more tests on this area? If so this could closed automatically by respective commit.

If I'm not wrong tests only check initialization - done is result of a previous report. OpenSSL itself does not use serialization, in particular case current IV. Is result tests does not cover regression in this area. With internal test PR will not pass regression test.
Personally I did not plan to test any more with master until official release.


Regards,
Roumen Petrov


P.S.
Every time when I thy tests fail. Last time was with an engine where creation of pkcs12 files fail. This prevent my to test quite simple functionality and also will fail in PKIX-SSH where regression tests produce pkcs12 files for third software.

I prefer to work on modernisation on PKIX-SSH - to minimize direct use of RSA, DSA and etc. keys and to use only PKEY. It was started long time ago and now is almost finished for identities(keys) - still used in key "creation" but I guess that this could be changed easily to "provider" like interface.
But there is number of other functionality that still needed to be changed. Work on key exchange is not even stated.

Due to current state of master (not functional in last 1.5 years) I have to change my plans. Support for providers is planed for next major OpenSSL release (3.1 or 4.0)
 From my point of view is too late to support 3.0.
Now my plans include only to minimize "deprecation" warnings with code refactor and to replace with exiting not yet deprecated functionality.

**Comment by t8m** [comment]
Unfortunately it is not clear to me, how the additional test-case should look like. There is the test_evp_iv() in evp_extra_test.c that is supposed to check that the running (updated) IV parameter is properly updated after the encryption call. Isn't that sufficient?

**Comment by petrovr** [comment]
Hi,

Tomáš Mráz wrote:
> Unfortunately it is not clear to me, how the additional test-case should look like. There is the test_evp_iv() in evp_extra_test.c that is supposed to check that the running (updated) IV parameter is properly updated after the encryption call. Isn't that sufficient?
> 
> 

I see - test_evp_iv() it is quite new :
---
 From 440b852a0f79ac4a7e101606a7c32d212e06d203 Mon Sep 17 00:00:00 2001
From: Benjamin Kaduk <bkaduk@akamai.com>
Date: Fri, 19 Jun 2020 20:42:29 -0700
Subject: [PATCH] Add tests for new EVP_CIPHER_CTX IV accessors
---

Quick review shows that test is limited. It does not test all ciphers.


In files like evpciph_*.txt IV is listed but content lack "Next IV". Is it not "current"/"updated" IV?


Regards
Roumen Petrov

**Comment by t8m** [comment]
Yes, but it at least tests all the relevant modes of aes-128. The evp_test that uses the evpciph_*.txt does not support checking the updated IV. I've opened a new issue #14047 for that as this issue is quite convoluted and IMO the base problem is now resolved.
